src/duckling/game/drawable/animated-drawable.component.ts

```diff
@@ -0,0 +1,109 @@
+import {
+    Component,
+    Input,
+    Output,
+    EventEmitter,
+    forwardRef,
+    AfterViewInit
+} from '@angular/core';
+import {MD_CARD_DIRECTIVES} from '@angular2-material/card';
+import {MD_LIST_DIRECTIVES} from '@angular2-material/list';
+
+import {Accordian, FormLabel, EnumChoiceComponent, NumberInput} from '../../controls';
+import {immutableAssign, immutableArrayAssign} from '../../util';
+
+
+import {AnimatedDrawable} from './animated-drawable';
+import {getDefaultDrawable, DrawableComponent} from './drawable.component';
+import {Drawable, DrawableType, drawableTypeToCppType} from './drawable';
+
+/**
+ * Component used to edit an Aniamted Drawable including all its children drawables
+ */
+@Component({
+    selector: "dk-animated-drawable-component",
+    styleUrls: ['./duckling/game/drawable/animated-drawable.component.css'],
+    directives: [
+        MD_CARD_DIRECTIVES,
+        Accordian,
+        FormLabel,
+        EnumChoiceComponent,
+        NumberInput,
+        forwardRef(() => DrawableComponent)],
+    template: `
+        <dk-number-input
+            label="Duration (seconds)"
+            [value]="animatedDrawable.duration"
+            (validInput)="onDurationChanged($event)">
+        </dk-number-input>
+
+        <dk-form-label title="Add Frame"></dk-form-label>
+        <dk-enum-choice
+            [enum]="DrawableType"
+            [selected]="DrawableType.Shape"
+            (addClicked)="onNewDrawableClicked($event)">
+        </dk-enum-choice>
+
+        <md-card
+            *ngIf="animatedDrawable?.frames?.length > 0"
+            class="drawables-card">
+            <dk-accordian
+                [elements]="animatedDrawable?.frames"
+                titleProperty="key"
+                keyProperty="key"
+                (elementDeleted)="onChildDrawablesChanged($event)"
+                (elementMovedDown)="onChildDrawablesChanged($event)"
+                (elementMovedUp)="onChildDrawablesChanged($event)">
+                <template let-element="$element" let-index="$index">
+                    <dk-drawable-component
+                        [drawable]="element"
+                        (drawableChanged)="onChildDrawableChanged(index, $event)">
+                    </dk-drawable-component>
+                </template>
+            </dk-accordian>
+        </md-card>
+    `
+})
+export class AnimatedDrawableComponent {
+    // hoist DrawableType so template can access it
+    DrawableType = DrawableType;
+
+    @Input() animatedDrawable : AnimatedDrawable;
+    @Output() drawableChanged = new EventEmitter<AnimatedDrawable>();
+
+    onChildDrawableChanged(index : number, newDrawable : Drawable) {
+        let head = this.animatedDrawable.frames.slice(0, index);
```

Comment:
It is more efficient to copy the entire array and then modify that one index in the copy (also less code).